x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/action/TransportMonitoringMigrateAlertsAction.java

```diff
@@ -0,0 +1,246 @@
+import org.elasticsearch.xpack.monitoring.exporter.Exporter.ExporterResourceStatus;
+import org.elasticsearch.xpack.monitoring.exporter.Exporters;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+public class TransportMonitoringMigrateAlertsAction extends TransportMasterNodeAction<MonitoringMigrateAlertsRequest,
+    MonitoringMigrateAlertsResponse> {
+
+    private static final Logger logger = LogManager.getLogger(TransportMonitoringMigrateAlertsAction.class);
+
+    private static final TimeValue ALERT_MIGRATION_BACK_OFF_RETRY = new TimeValue(5, TimeUnit.SECONDS);
+    private static final int ALERT_MIGRATION_MAX_RETRY = 3;
+
+    private final Client client;
+    private final Exporters exporters;
+
+    @Inject
+    public TransportMonitoringMigrateAlertsAction(Client client, Exporters exporters, TransportService transportService,
+                                                  ClusterService clusterService, ThreadPool threadPool,
+                                                  ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(MonitoringMigrateAlertsAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            MonitoringMigrateAlertsRequest::new, indexNameExpressionResolver, MonitoringMigrateAlertsResponse::new,
+            ThreadPool.Names.MANAGEMENT);
+        this.client = client;
+        this.exporters = exporters;
+    }
+
+    @Override
+    protected void masterOperation(Task task, MonitoringMigrateAlertsRequest request, ClusterState state,
+                                   ActionListener<MonitoringMigrateAlertsResponse> listener) throws Exception {
+        Settings.Builder decommissionAlertSetting = Settings.builder().put(Monitoring.MIGRATION_DECOMMISSION_ALERTS.getKey(), true);
+        client.admin().cluster().prepareUpdateSettings().setPersistentSettings(decommissionAlertSetting)
+            .execute(afterSettingUpdate(listener));
+    }
+
+    /**
+     * Executed after the settings update has been accepted, this collects the enabled and disabled exporters, requesting each of them
+     * to explicitly remove their installed alerts if possible. This makes sure that alerts are removed in a timely fashion instead of
+     * waiting for metrics to be bulked into the monitoring cluster.
+     */
+    private ActionListener<ClusterUpdateSettingsResponse> afterSettingUpdate(ActionListener<MonitoringMigrateAlertsResponse> listener) {
+        return ActionListener.wrap(clusterUpdateSettingsResponse -> {
+
+            // Ensure positive result
+            if (!clusterUpdateSettingsResponse.isAcknowledged()) {
+                listener.onFailure(new ElasticsearchException("Failed to update monitoring migration settings"));
```

Comment:
nit: include the setting name in the error.
nit: prefer `== false` over `!`

Reply:
If the setting is already true, does clusterUpdateSettingResponse.isAcknoledged() still return true ?

Reply:
I added a test that runs the migration multiple times. No failures encountered.